docs/core_classes.markdown

```diff
@@ -0,0 +1,228 @@
+# Core Classes Schema
+
+![Core Classes Diagram](assets/images/classes_schema.png)
+
+All objects are derived from the Pydantic `BaseModel` class, which provides data validation and serialization capabilities.
+
+## Action and Observation abstractions
+Tool is the synonym of function that has its signature containing name, decsription and arguments, and could be executed.  
+We need 1 class to represent tool to the LLM and 1 class to represent the specific tool call produced by the LLM.  
+Also the tool is the synonym of the action in the RL sense - something that produced by the agent and changes state of the environment.  
+Our definitions based on the defacto standards for tool calling from OAI and LiteLLM implementation:  
+https://platform.openai.com/docs/guides/function-calling  
+https://docs.litellm.ai/docs/completion/function_call
+```
+class Tool(BaseModel):
+    """
+    Represents a tool specification with a type, name, description and arguments.
+    Compatible with OAI, Anthropic and VLLM definitions.
+
+    Attributes:
+        type (Literal["function"]): The type of the tool, which is always "function".
+        name (str): The name of the function.
+        description (str): A brief description of the function.
+        parameters (dict): A dictionary containing the parameters of the function.
+    """
+
+    type: Literal["function"] = "function"
+    name: str
+    description: str
+    parameters: dict
+
+    @classmethod
+    """create tool object from python function"""
+    def from_function(cls, func: Callable) -> "Tool":
+        pass
+
+    @property
+    def schema(self) -> Dict[str, Any]:
+    """produce dict that could be passed as tool schema into LLM api"""
+        return self.json_dump()
+
+class ToolCall(BaseModel):
+    """
+    A class representing a function call.
+
+    Attributes:
+        id (str): The identifier for the tool call.
+        name (str): The name of the function being called.
+        arguments (Any): The arguments to be passed to the function.
+    """
+    id: str
+    name: str
+    arguments: Dict[str, Any]
+
+class Content(BaseModel):
+    """Represents a piece of content in an observation."""
+    type: str  # e.g., "text/plain", "image/png"
+    data: Any  # The actual content data
+
+class Observation(BaseModel):
+    tool_call_id: str | None = None # first observation may not be linked to any tool call
+    contents: list[Content]
+    metadata: dict = Field(default_factory=dict)
+    reward_info: dict = Field(default_factory=dict)
+
+    def to_messages(self) -> List[LLMMessage]:
```

Comment:
I think this should just loop over the content and call to_message on the different content.